pkg/status/status.go

```diff
@@ -1,65 +1,89 @@
 package status
 
 import (
-	"context"
+	"fmt"
+	"sync"
+	"time"
 
 	configv1 "github.com/openshift/api/config/v1"
 	configclient "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
 	cohelpers "github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
 	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
 	log "github.com/sirupsen/logrus"
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
-const clusterOperatorName = "marketplace-operator"
+const (
+	clusterOperatorName     = "marketplace-operator"
+	minSyncsBeforeReporting = 4
+	// At this point, a failure is critical so all syncs must be successful
+	// to report a working operator
+	successRatio           = 1.0
+	maxSyncsBeforeTruncate = 10000
+	syncTruncateValue      = 100
+)
+
+// SyncCh is used to report a sync event
+// The ratio of Successful Syncs / Total Syncs will be used to report operator health to the CVO
+var syncCh = make(chan error)
+
+// status will be a singleton
+var instance *status
+var mu sync.Mutex
 
 type status struct {
 	configClient    *configclient.ConfigV1Client
 	coAPINotPresent bool
 	namespace       string
 	clusterOperator *configv1.ClusterOperator
 	version         string
+	successfulSyncs int
+	syncs           int
 }
 
-// Status is the interface to report the health of the operator to CVO
-type Status interface {
-	SetFailing(message string)
-	SetAvailable(message string)
-	SetProgressing(message string)
-	GetVersion() string
-}
-
-func (s *status) GetVersion() string {
-	return s.version
+// SendSyncMessage is used to send messages to the SyncCh
+func SendSyncMessage(err error) {
+	if instance != nil {
+		// If the coAPI is not available do not send messages to the sync channel
+		if !instance.coAPINotPresent {
+			syncCh <- nil
```

Comment:
I would recommend using `select` and `default` to make it a non blocking call ( even though the channel is buffered ).
```
select {
	case syncCh <- err:
	default:
	}
```